core/jvm/src/main/scala/scalaz/zio/RTS.scala

```diff
@@ -849,25 +805,28 @@ private object RTS {
     final def register(cb: Callback[E, A]): Async[E, A] =
       observe0 {
         case ExitResult.Completed(r) => cb(r)
-        case ExitResult.Failed(e, _) => e
-        case ExitResult.Terminated(ts) =>
-          rts.submit(rts.unsafeRun(unhandled(ts)))
-          cb(ExitResult.Terminated(ts))
-      }.fold(identity, ExitResult.Failed(_, _), ExitResult.Terminated(_))
+        case ExitResult.Failed(cause) =>
+          cause.checkedFirst match {
+            case Some(_) =>
+            case None =>
+              rts.submit(rts.unsafeRun(unhandled(cause.toThrowable :: Nil)))
```

Comment:
I don't know why the `unhandled` exception handler is being invoked during `register`. Seems likely wrong. I'd delete that line. The place to do that is `done`.

@regiskuckaertz Can you take a look? This seems really strange to me.